File: backend/app/optical/registry.py
"""Kind Registry — Python mirror of frontend/src/optical/registry.ts.

Each op is a callable: (BeamRay, PhysicsOpContext) -> list[BeamRay].
Pure functions, no I/O, no side effects.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Optional, Union

from .beam_ray import BeamRay, Vec3


# ---------------------------------------------------------------------------
# Public types
# ---------------------------------------------------------------------------

# OpticalKind taxonomy (string aliases for clarity)
OpticalKind = str  # one of {"lens", "mirror", "polarizer", ...}


@dataclass(frozen=True)
class Face:
    id: str
    position_mm_body_local: Vec3
    normal_body_local: Optional[Vec3] = None
    aperture_mm: float = 0.0
    aperture_shape: str = "rectangle"   # "rectangle" | "ellipse" | "circle"


# Transfer matrix: 2x2, separate xy, or full 5x5. Each as tuple-of-tuples
# (immutable) so it can be hashed / compared.
TransferMatrix = Union[
    tuple[str, tuple[tuple[float, ...], ...]],   # ("abcd", ((A,B),(C,D)))
    tuple[str, tuple[tuple[float, ...], ...], tuple[tuple[float, ...], ...]],  # ("abcdXY", Mx, My)
    tuple[str, tuple[float, ...]],                # ("matrix5x5", (25 floats,))
]


KindParams = dict[str, object]
DynamicSources = dict[str, object]


@dataclass(frozen=True)
class PhysicsOpContext:
    face_in: Face
    face_out: Face
    params: KindParams
    dynamic: Optional[DynamicSources] = None
    transfer_matrix: Optional[TransferMatrix] = None
    # Internal face chain for multi-hop reflective elements (see
    # asset-physics-model.md §3.3). Empty tuple = 2-port slab.
    # Tracer applies mirror at B*-prefixed faces, Snell at A*-prefixed.
    face_via: tuple[Face, ...] = ()


PhysicsOp = Callable[[BeamRay, PhysicsOpContext], list[BeamRay]]


@dataclass
class KindEntry:
    ops: dict[str, PhysicsOp] = field(default_factory=dict)
    needs_aperture: bool = True
    default_wavelength_range_nm: Optional[tuple[float, float]] = None


# ---------------------------------------------------------------------------
# Registry
# ---------------------------------------------------------------------------

_REGISTRY: dict[OpticalKind, KindEntry] = {}


def register_kind(kind: OpticalKind, entry: KindEntry) -> None:
    if kind in _REGISTRY:
        raise ValueError(f'kind "{kind}" already registered')
    _REGISTRY[kind] = entry


def register_ops(kind: OpticalKind, ops: dict[str, PhysicsOp]) -> None:
    """Add additional ops to an already-registered kind. Used by variant
    modules (e.g. glan-laser adds glan_transmit_p / glan_reject_s under
    the existing polarizer kind)."""
    entry = _REGISTRY.get(kind)
    if entry is None:
        raise KeyError(
            f'register_ops: kind "{kind}" not registered yet — '
            f'import its base module first'
        )
    for name, op in ops.items():
        if name in entry.ops:
            raise ValueError(
                f'register_ops: op "{name}" already registered under kind "{kind}"'
            )
        entry.ops[name] = op


def get_op(kind: OpticalKind, op_name: str) -> PhysicsOp:
    entry = _REGISTRY.get(kind)
    if entry is None:
        raise KeyError(f'kind "{kind}" not registered')
    op = entry.ops.get(op_name)
    if op is None:
        raise KeyError(f'op "{op_name}" not found in kind "{kind}"')
    return op


def has_op(kind: OpticalKind, op_name: str) -> bool:
    entry = _REGISTRY.get(kind)
    return entry is not None and op_name in entry.ops


def list_registered_kinds() -> list[OpticalKind]:
    return list(_REGISTRY.keys())


def _clear_registry_for_tests() -> None:
    _REGISTRY.clear()

File: frontend/src/optical/registry.ts
/**
 * Kind Registry — central catalog of PhysicsOps and per-kind metadata.
 *
 * The registry lives in code (not DB): every PhysicsOp implementation
 * is registered here by name, and Asset3D records reference ops by
 * string. This decouples Asset3D data (which faces, which transitions,
 * which op name) from the actual op implementation (TypeScript today,
 * Rust+WASM in Phase 5+).
 *
 * Each op is a pure function: (BeamRay, ctx) → BeamRay[]. The ray
 * tracer is responsible for face-hit detection, frame transforms, and
 * dispatching to the right op based on `transition.op` string.
 */

import type { BeamRay, Vec3 } from "./beam-ray";

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** Top-level optical kind taxonomy. Extends as new kinds are added. */
export type OpticalKind =
  | "laser_source"
  | "tapered_amplifier"
  | "lens"
  | "mirror"
  | "dichroic_mirror"
  | "polarizer"
  | "waveplate"
  | "beam_splitter"
  | "pbs"
  | "aom"
  | "eom"
  | "faraday_rotator"
  | "fiber_coupler"
  | "fiber"
  | "fiber_end"
  | "isolator"
  | "nonlinear_crystal"
  | "saturable_absorber"
  | "detector"
  | "camera"
  | "spectrometer"
  | "wavemeter"
  | "beam_dump";

/** Face descriptor (from Asset3D.faces). Subset needed by ops. */
export type Face = {
  id: string;
  positionMmBodyLocal: Vec3;
  normalBodyLocal?: Vec3;
  apertureMm: number;
  apertureShape: "rectangle" | "ellipse" | "circle";
};

/** Geometric transfer matrix attached to a transition. Three forms:
 *   - `abcd`:    2×2 (applies equally to qx and qy)
 *   - `abcdXY`:  separate 2×2 for x and y axes (astigmatic elements)
 *   - `matrix5x5`: full 5×5 (with absolute-offset E_x/E_y term)
 *  The op chooses which form to consume. */
export type TransferMatrix =
  | { kind: "abcd"; M: [[number, number], [number, number]] }
  | { kind: "abcdXY";
      Mx: [[number, number], [number, number]];
      My: [[number, number], [number, number]];
    }
  | { kind: "matrix5x5"; M: number[] /* row-major flat, length 25 */ };

/** Free-form parameter bag (merged from asset.defaultParams ⊕ dynamicSources). */
export type KindParams = Record<string, unknown>;

/** Dynamic per-instance state from SceneObject (laser power, AOM freq, ...). */
export type DynamicSources = Record<string, unknown>;

/** Context passed to every PhysicsOp call. */
export type PhysicsOpContext = {
  faceIn: Face;
  faceOut: Face;
  params: KindParams;
  dynamic?: DynamicSources;
  transferMatrix?: TransferMatrix;
  /** Internal face chain for multi-hop reflective elements (see
   *  asset-physics-model.md §3.3). Tracer applies mirror at B*-prefixed
   *  faces, Snell at A*-prefixed. Empty/undefined = 2-port slab. */
  faceVia?: Face[];
};

/** A PhysicsOp transforms one input ray into zero or more output rays.
 *  Pure function (no side effects, no DOM, no THREE). */
export type PhysicsOp = (rayIn: BeamRay, ctx: PhysicsOpContext) => BeamRay[];

/** Per-kind metadata registered alongside ops. */
export type KindEntry = {
  ops: Record<string, PhysicsOp>;     // op name → impl
  needsAperture: boolean;
  defaultWavelengthRangeNm?: [number, number];
};

// ---------------------------------------------------------------------------
// Registry
// ---------------------------------------------------------------------------

const REGISTRY: Partial<Record<OpticalKind, KindEntry>> = {};

/** Register a kind's metadata and ops. Called once per kind module at
 *  import time (e.g. `kinds/lens/physics.ts` calls `registerKind("lens", ...)`). */
export function registerKind(kind: OpticalKind, entry: KindEntry): void {
  if (REGISTRY[kind]) {
    throw new Error(`kind "${kind}" already registered`);
  }
  REGISTRY[kind] = entry;
}

/** Add additional ops to an already-registered kind without re-declaring
 *  it. Used by variant modules (e.g. `kinds/glan-laser` adds glan_transmit_p
 *  / glan_reject_s under the existing `polarizer` kind). */
export function registerOps(kind: OpticalKind, ops: Record<string, PhysicsOp>): void {
  const entry = REGISTRY[kind];
  if (!entry) {
    throw new Error(
      `registerOps: kind "${kind}" not registered yet — import its base module first`,
    );
  }
  for (const [name, op] of Object.entries(ops)) {
    if (name in entry.ops) {
      throw new Error(
        `registerOps: op "${name}" already registered under kind "${kind}"`,
      );
    }
    entry.ops[name] = op;
  }
}

/** Look up a PhysicsOp by kind + op name. Throws if not found —
 *  callers should validate Asset3D.transitions[].op against registry
 *  at load time, not at trace time. */
export function getOp(kind: OpticalKind, opName: string): PhysicsOp {
  const entry = REGISTRY[kind];
  if (!entry) throw new Error(`kind "${kind}" not registered`);
  const op = entry.ops[opName];
  if (!op) throw new Error(`op "${opName}" not found in kind "${kind}"`);
  return op;
}

/** Predicate version of `getOp` — useful for validation. */
export function hasOp(kind: OpticalKind, opName: string): boolean {
  const entry = REGISTRY[kind];
  return !!entry && opName in entry.ops;
}

/** List all registered kinds (debug / introspection). */
export function listRegisteredKinds(): OpticalKind[] {
  return Object.keys(REGISTRY) as OpticalKind[];
}

/** Test-only: clear registry. Avoid in production code. */
export function _clearRegistryForTests(): void {
  for (const k of Object.keys(REGISTRY)) {
    delete REGISTRY[k as OpticalKind];
  }
}
